agnostic-svelte-ts/src/lib/components/Tabs/Tabs.svelte

```diff
@@ -0,0 +1,400 @@
+		border-top-right-radius: var(--agnostic-tabs-radius, 0.25rem);
+	}
+
+	.tab-item.tab-button.active {
+		color: var(--agnostic-gray-dark);
+
+		/* TODO -- update these colors to use css properties */
+		background-color: #fff;
+		border-color: var(--agnostic-gray-light) var(--agnostic-gray-light) #fff;
+	}
+
+	.tab-item:hover,
+	.tab-button:focus {
+		border-color: var(--agnostic-focus-ring-outline-width)
+			var(--agnostic-focus-ring-outline-width) var(--agnostic-gray-light);
+		isolation: isolate;
+		cursor: pointer;
+	}
+
+	.tabs-vertical .tab-button {
+		border: none;
+	}
+
+	.tab-button:disabled {
+		color: var(--agnostic-tabs-disabled-bg, var(--agnostic-gray-mid-dark));
+		background-color: transparent;
+		border-color: transparent;
+		opacity: 80%;
+	}
+
+	/**
+ * Elects to additively use the AgnosticUI custom focus ring alongside the border
+ * we already add above. It just makes things look more consistent across components.
+ * For example, when we tab into the panels and links within.
+ */
+	.tab-button-base:focus,
+	.tab-panel:focus,
+	.tab-button:focus {
+		box-shadow: 0 0 0 var(--agnostic-focus-ring-outline-width)
+			var(--agnostic-focus-ring-color);
+
+		/* Needed for High Contrast mode */
+		outline: var(--agnostic-focus-ring-outline-width)
+			var(--agnostic-focus-ring-outline-style)
+			var(--agnostic-focus-ring-outline-color);
+		transition: box-shadow var(--agnostic-timing-fast) ease-out;
+	}
+
+	@media (prefers-reduced-motion), (update: slow) {
+		.tab-button,
+		.tab-button-base:focus,
+		.tab-button:focus,
+		.tab-panel:focus,
+		.tab-list,
+		.tab-skinned {
+			transition-duration: 0.001ms !important;
+		}
+	}
+</style>
+
+<script lang="ts">
+	import type { NavigationDirection, Tab, TabSizes } from './api';
+
+	export let size: TabSizes = '';
+
+	/**
+   * This is an array of objects in shape like:
+   * {
+      title: "Tab 3",
+      tabPanelComponent: MyTabPanel,
+      // This is optional and only if you want to supply your own custom tab buttons.
+      // If not passed, we will generate an internal default tab button.
+      tabButtonComponent: TabButton3,
+    },
+   */
+	export let tabs: Tab[] = [];
+
+	export let isBorderless = false;
+	export let isVerticalOrientation = false;
+	export let isDisabled = false;
+	export let disabledOptions: string[] = [];
+	export let isSkinned = true;
+
+	/**
+	 * Explanation: we have two ways that the tab buttons get created:
+	 * 1. The `tabs` input array has dynamic `tabButtonComponent` components.
+	 * 2. The `tabs` has no `tabButtonComponent` and so we generate the tab
+	 * button internally.
+	 *
+	 * As such, the `dynamicComponentRefs` below are refs for case 1. and
+	 * `tabButtonRefs` are refs for case 2.
+	 */
+	let dynamicComponentRefs = []; // https://svelte.dev/tutorial/component-this
+	let tabButtonRefs = [];
+
+	// handle element removal by filtering null
+	$: dynamicComponentRefs = [];
+	// dynamicComponentRefs.filter(el => el);
```

Comment:
What a mess I left 🤦‍♂️

Reply:
No worries, the comments actually helped me understand the code a bit quicker! 😄 

I'm still thinking about refactoring comments as a follow-up:
- colocate it with the code (eg. move this explanation to where the filtering now happens)
- reformat comments on exported values so that the description looks good both in IntelliSense and Storybook
  eg. tabs prop:  
  | IntelliSense | Storybook |
  | ------ | ----- |
  | <img alt="image" src="https://user-images.githubusercontent.com/11555574/168245756-0538e7ce-3ce9-4c1b-983d-4ff194c4237f.png"> | <img alt="image" src="https://user-images.githubusercontent.com/11555574/168245635-43b7407c-7998-4ab4-a0bc-56d4ed1dc4c3.png"> |
- perhaps remove some type hint comments that are redundant with the introduction of TS